File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/adapter/lab/AddressBook.m
% AddressBook.m

classdef AddressBook < handle & DpContainer
	properties (GetAccess = private, SetAccess = private)
		List % DpList of Person
	end

	methods
		% constructor:
		function theAddressBook = AddressBook
			theAddressBook.List = DpList;
		end

		function add(theAddressBook, aPerson)
			if isempty(aPerson)
				% throw an exception.
			end

			theAddressBook.List.add(aPerson);
		end

		function result = getNumberOfPersons(theAddressBook)
			result = theAddressBook.List.getSize;
		end

		function remove(theAddressBook, aPerson)
			% I was too lazy to implement this. :-(
		end

		function result = makeIterator(theAddressBook)
			result = theAddressBook.List.makeIterator;
		end

		function result = getPerson(theAddressBook, which)
			aIterator = theAddressBook.makeIterator;
			aCounter = 1;
			while ~aIterator.isDone
				if aCounter == which
					result = aIterator.getCurrent;
					return;
				end
				aIterator.advance;
				aCounter = aCounter + 1;
			end
			result = Person.empty;
		end
	end
end

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/adapter/solution/AddressBook.java
// AddressBook.java

public class AddressBook extends DpContainer<Person>
{
	public AddressBook()
	{
		myPersons = new DpList<Person>();
	}

	public void add(Person aPerson)
	{
		if (aPerson == null)
		{
			// throw an exception.
		}

		myPersons.add(aPerson);
	}

	public int getNumberOfPersons()
	{
		return myPersons.getSize();
	}

	public void remove(Person aPerson)
	{
		// I was too lazy to implement this. :-(
	}

	public DpIterator<Person> makeIterator()
	{
		return myPersons.makeIterator();
	}

	public Person getPerson(int which)
	{
		DpIterator<Person> aIterator = makeIterator();
		int aCounter = 0;
		while (!aIterator.isDone())
		{
			if (aCounter == which)
			{
				return aIterator.getCurrent();
			}
			aIterator.advance();
			++aCounter;
		}
		return null;
	}

	private DpList<Person> myPersons;
}
